bin/publish.js

```diff
@@ -0,0 +1,327 @@
+  let status = execWithLog('git status');
+
+  if (!status.match(/^nothing to commit/m)) {
+    if (options.force) {
+      console.log(
+        chalk.white('⚠️ ⚠️ ⚠️  Local Git branch has uncommitted changes!\n\t') +
+          chalk.yellow('Passed option: ') +
+          chalk.white('--force') +
+          chalk.grey(' :: ignoring unclean git working tree')
+      );
+    } else {
+      console.log(
+        chalk.red('💥 Git working tree is not clean. 💥 \n\t') +
+          chalk.grey('Use ') +
+          chalk.white('--force') +
+          chalk.grey(' to ignore this warning and publish anyway\n') +
+          chalk.yellow(
+            '⚠️  Publishing from an unclean working state may result in a broken release ⚠️'
+          )
+      );
+      process.exit(1);
+    }
+  }
+
+  if (!status.match(/^Your branch is up to date with/m)) {
+    if (options.force) {
+      console.log(
+        chalk.white('⚠️ ⚠️ ⚠️  Local Git branch is not in sync with origin branch') +
+          chalk.yellow('\n\tPassed option: ') +
+          chalk.white('--force') +
+          chalk.grey(' :: ignoring unsynced git branch')
+      );
+    } else {
+      console.log(
+        chalk.red('💥 Local Git branch is not in sync with origin branch. 💥 \n\t') +
+          chalk.grey('Use ') +
+          chalk.white('--force') +
+          chalk.grey(' to ignore this warning and publish anyway\n') +
+          chalk.yellow(
+            '⚠️  Publishing from an unsynced working state may result in a broken release ⚠️'
+          )
+      );
+      process.exit(1);
+    }
+  }
+
+  let expectedChannelBranch = options.distTag === 'canary' ? 'master' : options.distTag;
+
+  if (options.channel === 'lts') {
+    expectedChannelBranch = `lts-${semver.major(options.currentVersion)}-${semver.minor(
+      options.currentVersion
+    )}`;
+  }
+
+  let foundBranch = status.split('\n')[0];
+  foundBranch = foundBranch.replace('On branch ', '');
+
+  if (foundBranch !== expectedChannelBranch) {
+    if (options.force) {
+      console.log(
+        chalk.white(
+          `⚠️ ⚠️ ⚠️  Expected to publish npm tag ${
+            options.distTag
+          } from the git branch ${expectedChannelBranch}, but found ${foundBranch}`
+        ) +
+          chalk.yellow('\n\tPassed option: ') +
+          chalk.white('--force') +
+          chalk.grey(' :: ignoring unexpected branch')
+      );
+    } else {
+      console.log(
+        chalk.red(
+          `💥 Expected to publish npm tag ${
+            options.distTag
+          } from the git branch ${expectedChannelBranch}, but found ${foundBranch} 💥 \n\t`
+        ) +
+          chalk.grey('Use ') +
+          chalk.white('--force') +
+          chalk.grey(' to ignore this warning and publish anyway\n') +
+          chalk.yellow('⚠️  Publishing from an incorrect branch may result in a broken release ⚠️')
+      );
+      process.exit(1);
+    }
+  }
+}
+
+function retrieveNextVersion() {
+  /* 
+  #master lerna.json 3.11.0-canary.x
```

Comment:
I am having a bit of trouble understanding what this comment means

Reply:
It's a brief rundown of how version updates flow through the branches. Basically we only ever bump the major or minor version on `master`, and all other branches pick it up as those changes flow through the release cycle.